# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/windows/msaa/EnumVariant.cpp
# Commit: 1a948c0ec982
# Full Hash: 1a948c0ec9826d73bba860699f540cbb25f34e34
# Author: James Teh <jteh@mozilla.com>
# Date: 2021-04-29 21:42:31
# Regressor Bug: 1694865
# File Overlap Count: 1
# Description:
#   Bug 1694865 part 18: Make EnumVariant, GeckoCustom and sdn*Accessible classes aggregate MsaaAccessible/MsaaDocAccessible instead of AccessibleWrap/DocAccessibleWrap. r=morgan
#   
#   These objects need to aggregate the primary COM object (MsaaAccessible).
#   Once the IUnknown implementation moves out of AccessibleWrap, it won't be possible to aggregate *AccessibleWrap any more.
#   
# ==============================================================================

diff -r 3ea7b46bb4f8 -r 1a948c0ec982 accessible/windows/msaa/EnumVariant.cpp
--- a/accessible/windows/msaa/EnumVariant.cpp	Thu Apr 29 10:20:56 2021 +0000
+++ b/accessible/windows/msaa/EnumVariant.cpp	Thu Apr 29 10:20:57 2021 +0000
@@ -15,7 +15,7 @@
 
 IMPL_IUNKNOWN_QUERY_HEAD(ChildrenEnumVariant)
 IMPL_IUNKNOWN_QUERY_IFACE(IEnumVARIANT)
-IMPL_IUNKNOWN_QUERY_TAIL_AGGREGATED(mAnchorAcc)
+IMPL_IUNKNOWN_QUERY_TAIL_AGGREGATED(mAnchorMsaa)
 
 STDMETHODIMP
 ChildrenEnumVariant::Next(ULONG aCount, VARIANT FAR* aItems,
@@ -24,7 +24,8 @@
 
   *aCountFetched = 0;
 
-  if (mAnchorAcc->IsDefunct() || mAnchorAcc->LocalChildAt(mCurIndex) != mCurAcc)
+  AccessibleWrap* anchor = mAnchorMsaa->LocalAcc();
+  if (!anchor || anchor->LocalChildAt(mCurIndex) != mCurAcc)
     return CO_E_OBJNOTCONNECTED;
 
   ULONG countFetched = 0;
@@ -34,7 +35,7 @@
     IDispatch* accNative = MsaaAccessible::NativeAccessible(mCurAcc);
 
     ++mCurIndex;
-    mCurAcc = mAnchorAcc->LocalChildAt(mCurIndex);
+    mCurAcc = anchor->LocalChildAt(mCurIndex);
 
     // Don't output the accessible and count it as having been fetched unless
     // it is non-null
@@ -55,21 +56,23 @@
 
 STDMETHODIMP
 ChildrenEnumVariant::Skip(ULONG aCount) {
-  if (mAnchorAcc->IsDefunct() || mAnchorAcc->LocalChildAt(mCurIndex) != mCurAcc)
+  AccessibleWrap* anchor = mAnchorMsaa->LocalAcc();
+  if (!anchor || anchor->LocalChildAt(mCurIndex) != mCurAcc)
     return CO_E_OBJNOTCONNECTED;
 
   mCurIndex += aCount;
-  mCurAcc = mAnchorAcc->LocalChildAt(mCurIndex);
+  mCurAcc = anchor->LocalChildAt(mCurIndex);
 
   return mCurAcc ? S_OK : S_FALSE;
 }
 
 STDMETHODIMP
 ChildrenEnumVariant::Reset() {
-  if (mAnchorAcc->IsDefunct()) return CO_E_OBJNOTCONNECTED;
+  AccessibleWrap* anchor = mAnchorMsaa->LocalAcc();
+  if (!anchor) return CO_E_OBJNOTCONNECTED;
 
   mCurIndex = 0;
-  mCurAcc = mAnchorAcc->LocalChildAt(0);
+  mCurAcc = anchor->LocalChildAt(0);
 
   return S_OK;
 }